### PR TITLE
Add VirtualHostName mapping feature

### LINKED ISSUE #154: Expose SetVirtualHostNameToFolderMapping of WebView2
API description: https://docs.microsoft.com/en-us/microsoft-edge/webview2/reference/win32/icorewebview2_3?view=webview2-1.0.1210.39#setvirtualhostnametofoldermapping

Reason: When browsing local HTML files that try to load local resources, we run into the notorious CORS error, as discussed prominently on [Stack Overflow](https://stackoverflow.com/questions/10752055/cross-origin-requests-are-only-supported-for-http-error-when-loading-a-local). We can spin up a fileserver in the app listening to localhost loopback requests, but doing so may be blocked on managed devices. We can add the --allow-file-access-from-files flag during initializeEnvironment(), but it seems this is ignored once we packaged our app as MSIX installer. Just to use the virtual host name folder mapping from Edge WebView2 would be the perfect workaround with minimal security and operational risk.

### PR DESCRIPTION
**Pull request by tlercher**
Also a rather simple implementation which fixes #154.

I only tested it with a simple `index.html` in a mapped folder, not the CORS-Feature-Flags itself.

### CODE REVIEW

File: lib/src/webview.dart [hunk 1]
```diff
@@ -566,3 +593,16 @@ class _WebviewState extends State<Webview> {
     _cursorSubscription?.cancel();
   }
 }
+
+extension WebviewHostResourceAccessKindExt on WebviewHostResourceAccessKind {
```
Comment (jnschulze) [inline lib/src/webview.dart:597]: For consistency reasons, we just pass the corresponding enum member's index via the method channel and have the same enum on the C++ side which is kept in sync with the Dart enum then. (see https://github.com/jnschulze/flutter-webview-windows/blob/91127a465bffeccf4d72d27f3c9ab8690a374449/windows/webview.h#L13-L29)

File: windows/webview.cc [hunk 2]
```diff
@@ -638,3 +638,32 @@ bool Webview::Resume() {
...
+  }
+
+  wil::com_ptr<ICoreWebView2_3> webview;
+  webview = webview_.query<ICoreWebView2_3>();
+  if (!webview) {
+    return false;
+  }
+
+  return webview->ClearVirtualHostNameToFolderMapping(towstring(hostName).c_str());
+}
```
Comment (jnschulze) [inline windows/webview.cc:669]: nit: Missing newline

**Review by jnschulze — CHANGES_REQUESTED** [review]

File: lib/src/webview.dart (hunk 1 above)
Comment (tlercher) [reply lib/src/webview.dart]: Good Point! I seem to completely overlooked that while glancing over the code.

I'll added it (and rearranged the order to match [`COREWEBVIEW2_HOST_RESOURCE_ACCESS_KIND`](https://docs.microsoft.com/en-us/microsoft-edge/webview2/reference/win32/icorewebview2#corewebview2_host_resource_access_kind).

Would be assigning the values from `COREWEBVIEW2_HOST_RESOURCE_ACCESS_KIND` to the new enum be a good option?

File: windows/webview.cc (hunk 2 above)
Comment (tlercher) [reply windows/webview.cc]: Done (if the trailing newline was meant)

File: lib/src/webview.dart (hunk 1 above)
Comment (jnschulze) [reply lib/src/webview.dart]: I would create an enum whose values are independent of the WebView2 enum values. I'd then cast the int received in the method channel handler in `webview_bridge.cc` to the internal enum `WebviewHostResourceAccessKind` (which needs to be kept in sync with your Dart enum). The `switch` mapping the internal enum to the WebView2 enum should be moved to the `Webview::SetVirtualHostNameMapping` implementation.

**Review by jnschulze — CHANGES_REQUESTED** [review]
Almost perfect. See my comments below.

File: windows/webview_bridge.cc [hunk 3]
```diff
@@ -422,6 +425,44 @@ void WebviewBridge::HandleMethodCall(
...
+    if (!list || list->size() != 3) {
+      return result->Error(kErrorInvalidArgs);
+    }
+
+    const auto hostName = std::get_if<std::string>(&(*list)[0]);
+    const auto path = std::get_if<std::string>(&(*list)[1]);
+    const auto accessKind = std::get_if<int32_t>(&(*list)[2]);
+
+    if (hostName && path && accessKind) {
+      switch (*accessKind) {
```
[the remark this answers is not in the document]
Comment (tlercher) [reply windows/webview_bridge.cc]: Done. But one last question to this topic (i hope), should there be a default value in dart for this (e.g. `deny`)?

File: lib/src/enums.dart [hunk 4]
```diff
@@ -26,3 +26,15 @@ enum WebviewPermissionDecision { none, allow, deny }
...
+/// The kind of cross origin resource access for virtual hosts
+///
+/// [deny] all cross origin requests are denied.
+/// [allow] all cross origin requests are allowed.
+/// [denyCors] sub resource cross origin requests are allowed, otherwise denied.
+///
+/// For more detailed information, please refer to
+/// [Microsofts](https://docs.microsoft.com/en-us/microsoft-edge/webview2/reference/win32/icorewebview2#corewebview2_host_resource_access_kind)
+/// documentation.
+// Order must match WebviewPointerButton (see webview.h)
```
[the remark this answers is not in the document]
Comment (tlercher) [reply lib/src/enums.dart]: Changed!

File: windows/webview_bridge.cc (hunk 3 above)
[the remark this answers is not in the document]
Comment (jnschulze) [reply windows/webview_bridge.cc]: For the method `addVirtualHostNameMapping` ?

[the remark this answers is not in the document]
Comment (tlercher) [reply windows/webview_bridge.cc]: Yep

[the remark this answers is not in the document]
Comment (jnschulze) [reply windows/webview_bridge.cc]: I'll leave the decision up to you

File: windows/webview_bridge.cc [hunk 5]
```diff
@@ -422,6 +425,35 @@ void WebviewBridge::HandleMethodCall(
...
+  if (method_name.compare(kMethodSetVirtualHostNameMapping) == 0) {
+    const flutter::EncodableList* list =
+    std::get_if<flutter::EncodableList>(method_call.arguments());
+    if (!list || list->size() != 3) {
+      return result->Error(kErrorInvalidArgs);
+    }
+
+    const auto hostName = std::get_if<std::string>(&(*list)[0]);
+    const auto path = std::get_if<std::string>(&(*list)[1]);
+    const auto accessKind = std::get_if<WebviewHostResourceAccessKind>(&(*list)[2]);
```
[the remark this answers is not in the document]
Comment (tlercher) [reply windows/webview_bridge.cc]: Yep, true! Changed.

File: windows/webview_bridge.cc (hunk 3 above)
[the remark this answers is not in the document]
Comment (tlercher) [reply windows/webview_bridge.cc]: Let's leave it as is then ;)

**Review by jnschulze — APPROVED** [review]
LGTM. Thanks fort your effort!

**Pull request merged**